checker/src/org/checkerframework/checker/upperbound/UpperBoundTransfer.java

```diff
@@ -0,0 +1,440 @@
+import org.checkerframework.dataflow.analysis.RegularTransferResult;
+import org.checkerframework.dataflow.analysis.TransferInput;
+import org.checkerframework.dataflow.analysis.TransferResult;
+import org.checkerframework.dataflow.cfg.node.ArrayCreationNode;
+import org.checkerframework.dataflow.cfg.node.AssignmentNode;
+import org.checkerframework.dataflow.cfg.node.EqualToNode;
+import org.checkerframework.dataflow.cfg.node.FieldAccessNode;
+import org.checkerframework.dataflow.cfg.node.GreaterThanNode;
+import org.checkerframework.dataflow.cfg.node.GreaterThanOrEqualNode;
+import org.checkerframework.dataflow.cfg.node.LessThanNode;
+import org.checkerframework.dataflow.cfg.node.LessThanOrEqualNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.NotEqualNode;
+import org.checkerframework.framework.flow.CFAbstractTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+public class UpperBoundTransfer
+        extends CFAbstractTransfer<CFValue, UpperBoundStore, UpperBoundTransfer> {
+
+    private UpperBoundAnnotatedTypeFactory atypeFactory;
+
+    private QualifierHierarchy qualifierHierarchy;
+
+    public UpperBoundTransfer(UpperBoundAnalysis analysis) {
+        super(analysis);
+        atypeFactory = (UpperBoundAnnotatedTypeFactory) analysis.getTypeFactory();
+        qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    }
+
+    // Refine the type of expressions used as an array dimension to be
+    // less than length of the array to which the new array is
+    // assigned.
+    @Override
+    public TransferResult<CFValue, UpperBoundStore> visitAssignment(
+            AssignmentNode node, TransferInput<CFValue, UpperBoundStore> in) {
+        AnnotationMirror UNKNOWN = atypeFactory.UNKNOWN;
+        TransferResult<CFValue, UpperBoundStore> result = super.visitAssignment(node, in);
+
+        // When an existing array is assigned into, the store needs to be cleared -
+        // that is, every instance of LTL and LTEL needs to be removed, since arrays
+        // might be aliased, and when an array is modified to be a different length,
+        // that could cause any of our information about arrays to be wrong.
+        if (node.getTarget().getType().getKind() == TypeKind.ARRAY) {
+            // This means that the existing store needs to be invalidated.
+            // As far as I can tell the easiest way to do this is to just
+            // create a new TransferResult.
+            TransferResult<CFValue, UpperBoundStore> newResult =
+                    new RegularTransferResult<CFValue, UpperBoundStore>(
+                            result.getResultValue(), new UpperBoundStore(analysis, true));
+            result = newResult;
+        }
+
+        // This handles when a new array is created.
```

Comment:
This comment isn't helpful.  How are new arrays handled?  Ok, now I see that comment on the method should be moved here.